fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,544 @@
+            // to Prometheus: number of keys, number of active nodes, etc.
+            let active_ingest_summary_node_mappings: Vec<&IngestSummaryNodeMapping> =
+                ingest_summary_node_mappings
+                    .iter()
+                    .filter(|ingest_summary_node_mapping| {
+                        ingest_summary_node_mapping.ingest_summary.mode
+                            == IngestControllerMode::Active
+                    })
+                    .collect();
+
+            let active_node_count = active_ingest_summary_node_mappings.len();
+            match active_node_count {
+                0 => {
+                    log::error!(
+                        self.logger,
+                        "There are currently no active nodes in the Fog Ingest cluster. Initiating automatic failover.",
+                    );
+                    match self.perform_automatic_failover(ingest_summary_node_mappings) {
+                        Ok(_) => {
+                            log::info!(self.logger, "Automatic failover completed successfully.")
+                        }
+                        Err(err) => {
+                            log::error!(self.logger, "Automatic failover failed: {:?}", err)
+                        }
+                    };
+                }
+                1 => {
+                    log::trace!(
+                        self.logger,
+                        "There is one active node in the Fog Ingest cluster. Active ingress key: {:?}",
+                        active_ingest_summary_node_mappings[0]
+                        .ingest_summary
+                        .get_ingress_pubkey()
+                    );
+                    continue;
+                }
+                _ => {
+                    let active_node_ingress_pubkeys: Vec<&external::CompressedRistretto> =
+                        active_ingest_summary_node_mappings
+                            .iter()
+                            .map(|active_ingest_summary_node_mapping| {
+                                active_ingest_summary_node_mapping
+                                    .ingest_summary
+                                    .get_ingress_pubkey()
+                            })
+                            .collect();
+                    log::warn!(
+                        self.logger,
+                        "There are multiple active nodes in the Fog Ingest cluster. Active ingress keys: {:?}",
+                        active_node_ingress_pubkeys
+                    );
+                    // TODO: Set up sentry alerts and signal to ops that two
+                    // keys are active at once. This is
+                    // unexpected.
+                }
+            }
+        }
+    }
+
+    /// Returns the latest round of ingest summaries for each
+    /// FogIngestGrpcClient that communicates with a node that is online.
+    fn retrieve_ingest_summary_node_mappings(&self) -> Vec<IngestSummaryNodeMapping> {
```

Comment:
I think this should return the error, so that the caller can abort this pass if some of the nodes are not reachable / not responding?

Reply:
if any of the nodes are not reachable then the overseer should not do anything, otherwise if the network partitions bad things can happen

Reply:
Yes good point. Originally, I was considering the case in which one of the nodes is continually unresponsive. In this scenario, there could be no active nodes and this unresponsive node could be an idle node. In this case, we'd want overseer to respond, but it wouldn't because of this one unresponsive node.

However, I agree with you that it's too dangerous to proceed with the logic if a node is unresponsive because it could be the active node and then come back online and there'd be two active nodes (a second one would be activated by overseer in the interim). 

I also think that because we're assuming that the unresponsive node will become responsive soon enough (i.e. Azure will bring the node back online in a relatively reasonable time frame), then the probability of this "infinite loop of inaction" is pretty low. 

I changed the method.